src/main/java/org/truffleruby/core/kernel/TruffleKernelNodes.java

```diff
@@ -92,6 +95,90 @@ public boolean load(VirtualFrame frame, DynamicObject file, boolean wrap,
+            if (getContext().getSharedObjects().isSharing()) {
+                writeBarrierNode.executeWriteBarrier(value);
+            }
+            storage.setValueInternal(value);
+            storage.updateAssumeConstant(getContext());
+            return value;
+        }
+
+        @Specialization(guards = { "isRubySymbol(cachedName)", "name == cachedName", "!storage.isAssumeConstant()" }, limit = "1", replaces = "writeAssumeConstant")
+        public Object write(DynamicObject name, Object value,
+                @Cached("name") DynamicObject cachedName,
+                @Cached("getStorage(name)") GlobalVariableStorage storage) {
+            if (getContext().getSharedObjects().isSharing()) {
+                writeBarrierNode.executeWriteBarrier(value);
+            }
+            storage.setValueInternal(value);
+            return value;
+        }
+
+        @Specialization(guards = "isRubySymbol(name)")
+        public Object writeGeneric(DynamicObject name, Object value) {
+            GlobalVariableStorage storage = getStorage(name);
+            if (getContext().getSharedObjects().isSharing()) {
+                writeBarrierNode.executeWriteBarrier(value);
+            }
+            storage.setValueInternal(value);
+            return value;
+        }
+
+        @TruffleBoundary
+        protected GlobalVariableStorage getStorage(DynamicObject name) {
+            return getContext().getCoreLibrary().getGlobalVariables().getStorage(Layouts.SYMBOL.getString(name));
+        }
+    }
+
+    @CoreMethod(names = "global_variable_get", isModuleFunction = true, required = 1)
+    public abstract static class ReadGlobalVariableNode extends CoreMethodArrayArgumentsNode {
+        @Specialization(guards = { "isRubySymbol(cachedName)", "name == cachedName" }, limit = "1", assumptions = "storage.getUnchangedAssumption()")
+        public Object readConstant(DynamicObject name,
+                @Cached("name") DynamicObject cachedName,
+                @Cached("getStorage(name)") GlobalVariableStorage storage,
+                @Cached("storage.getValue()") Object value) {
+            return value;
+        }
+
+        @Specialization(guards = { "isRubySymbol(cachedName)", "name == cachedName" }, limit = "1", replaces = "readConstant")
+        public Object read(DynamicObject name,
+                @Cached("name") DynamicObject cachedName,
+                @Cached("getStorage(name)") GlobalVariableStorage storage) {
+            return storage.getValue();
+        }
+
+        @Specialization(guards = "isRubySymbol(name)", replaces = "read")
+        public Object readGeneric(DynamicObject name) {
+            return getStorage(name).getValue();
+        }
+
+        @TruffleBoundary
+        protected GlobalVariableStorage getStorage(DynamicObject name) {
+            return getContext().getCoreLibrary().getGlobalVariables().getStorage(Layouts.SYMBOL.getString(name));
+        }
+    }
```

Comment:
This duplicates a lot of complex code.
Can the original code be removed in this PR as well or the original node reuse this (or vice versa)?